segmentstore/storage/impl/src/main/java/io/pravega/segmentstore/storage/impl/fs/FSStorage.java

```diff
@@ -0,0 +1,453 @@
+        retFuture.complete(exists);
+    }
+
+
+    private void syncCreate(String streamSegmentName, Duration timeout, CompletableFuture<SegmentProperties> retVal) {
+        log.info("Creating Segment {}", streamSegmentName);
+        try {
+            Set<PosixFilePermission> perms = new HashSet<>();
+            // add permission as rw-r--r-- 644
+            perms.add(PosixFilePermission.OWNER_WRITE);
+            perms.add(PosixFilePermission.OWNER_READ);
+            perms.add(PosixFilePermission.GROUP_READ);
+            perms.add(PosixFilePermission.OTHERS_READ);
+            FileAttribute<Set<PosixFilePermission>> fileAttributes = PosixFilePermissions.asFileAttribute(perms);
+
+            Path path = Paths.get(config.getNfsRoot(), streamSegmentName);
+            Files.createDirectories(path.getParent());
+            Files.createFile(path, fileAttributes);
+            log.info("Created Segment {}", streamSegmentName);
+            retVal.complete(this.getStreamSegmentInfo(streamSegmentName, timeout).get());
+        } catch (Exception e) {
+            log.info("Exception {} while creating a segment {}", e, streamSegmentName);
+            if (e instanceof FileAlreadyExistsException) {
+                retVal.completeExceptionally(new StreamSegmentExistsException(streamSegmentName, e));
+            } else {
+                retVal.completeExceptionally(e);
+            }
+        }
+    }
+
+    private void syncWrite(SegmentHandle handle, long offset, InputStream data, int length, Duration timeout,
+                           CompletableFuture<Void> retVal) {
+        log.trace("Writing {} to segment {} at offset {}", length, handle.getSegmentName(), offset);
+        Path path = Paths.get(config.getNfsRoot(), handle.getSegmentName());
+
+        if (handle.isReadOnly()) {
+            log.info("Write called on a readonly handle of segment {}", handle.getSegmentName());
+            retVal.completeExceptionally(new IllegalArgumentException());
+            return;
+        }
+
+        if (!Files.exists(path)) {
+            retVal.completeExceptionally(new StreamSegmentNotExistsException(handle.getSegmentName(), null));
+            return;
+        }
+
+        try {
+            if (!isWritableFile(path)) {
+                retVal.completeExceptionally(new StreamSegmentSealedException(handle.getSegmentName()));
+                return;
+            }
+
+            long fileSize = path.toFile().length();
+            if (fileSize < offset) {
+                retVal.completeExceptionally(new BadOffsetException(handle.getSegmentName(), fileSize, offset));
+            } else {
+                try (FileChannel channel = FileChannel.open(path,
+                        StandardOpenOption.WRITE); ReadableByteChannel sourceChannel = Channels.newChannel(data)) {
+                    long bytesWritten = channel.transferFrom(sourceChannel, offset, length);
+                    channel.force(true);
```

Comment:
I think you missed the second part of the question, why do we need to force the metadata sync as well? In mechanical drives, it typically induces another seek and sync.

I suspect that the metadata is necessary because we aren't pre-allocating space, so we might be adding blocks to the file and we clearly need to update the file metadata in this case. If we were pre-allocating, then we wouldn't need to force the metadata update. I'm not sure if it is worth pre-allocating in this case, but it might be worth thinking about it.